fvm/src/gas/price_list.rs

```diff
@@ -415,16 +509,19 @@ lazy_static! {
         get_randomness_base: Zero::zero(),
         get_randomness_per_byte: Zero::zero(),
 
-        block_memcpy_per_byte_cost: Gas::from_milligas(500),
-
-        block_open_base: Gas::new(114617),
-        block_open_memret_per_byte_cost: Gas::new(10),
+        block_memcpy_per_byte_cost: Gas::from_milligas(800),
+        block_create_memret_per_byte_cost: Gas::from_milligas(6400),
```

Comment:
memret is for memory "retention". I'm guessing that this per-byte cost needs to go somewhere else?

Reply:
For timing analysis, we should set this to 0.

Reply:
Thanks for commenting on it. I played around with the prices and eyeballed the formulas for a while to develop my own understanding of what retention might mean, based on the evidence. 

Here's what we started with IIRC:
* `on_block_read`: `block_read_base + block_memcpy_per_byte_cost * data_size`
* `on_block_create`: `block_create_base + (block_memcpy_per_byte_cost + block_create_memret_per_byte_cost) * data_size`

The value of `block_memcpy_per_byte_cost` was 500 and `block_create_memret_per_byte_cost` was 10. 

The first case is very simple with just one price in the formula, which is great because it determines `block_memcpy_per_byte_cost` unambiguously. That came out to be ~800. But then `on_block_create` was heavily under priced on the charts, so there had to be some other cost. Based on earlier experiments `Block::new` takes a long time when it comes from the slice of the memory passed to it, so I figured we can put the remaining of the cost into `block_create_memret_per_byte_cost`. Fore sure it is doing something different than `block_memcpy_per_byte_cost`. 

Then I looked at `on_block_open` which also creates a `Block`, but it does so from a `Vec<u8>` not a slice. What happens there is that the vector is adjusted to have 0 capacity, but there's no copying. 

Its formula was: `(block_open_memret_per_byte_cost + block_memcpy_per_byte_cost) * data_size`. I didn't actually see anything that looked comparable to `block_read` to justify `block_memcpy_per_byte_cost`, so I decided it might be best to remove this term from both `on_block_open` and `on_block_create`, and I thought we might want to live with the interpretation of "retention" as "create a `Block`", in which case the fact that in `block_create` we copy and in `block_open` we don't might be the reason why there are separate `open_memret` and `close_memret` variants. Otherwise I don't know why you would differentiate between the two. 

> For timing analysis, we should set this to 0.

Well, if you set it to zero, that means you treat it as a cost similar to the storage costs: people would have to pay just for data being in memory. And this memory is the host side block registry, not the Wasm memory, which they already paid for. 

That could make sense, although again, I don't know why `open` and `create` would have different prices in that case. It would potentially be more convenient to put it into the `storage_gas`. 

However, that would mean the formulas would be lacking some term to express the time cost. The current formula for `on_block_create` is `block_create_base + block_create_memret_per_byte_cost`. If we set the latter to 0, it becomes flat; I can't just put back `block_memcpy_per_byte_cost` because that is demonstrated to be cheaper by `on_block_read`. So we'd have to come up with something new. 

I actually tried this in https://github.com/filecoin-project/ref-fvm/pull/1291/commits/3a2317d79cef0d5fabaa1537bc423002988d4f59 but then I thought wait, there is a memret for open and close, maybe _that's_ why they are different. 

Maybe you can use these changes and the calibration methods in your PR to come up with new formulas, based on your deeper intuition about what's going on and what were the original intentions?

Reply:
Ah, got it. Ok, I'll try to fit that cost somewhere else.

Basically, the "memory retention" cost was purely a way to prevent us from running out of memory (which is why I combined them into a single charge in the other PR). Ideally, we'd just track memory of open blocks, but that's more complicated (and has been punted out till M2.2).

Reply:
Yeah, I thought it's something like that, it was just confusing to see it in `compute_gas`, which I thought relates to time. It would be awkward to have to remember to go back to prices to zero some costs out before we can run a time analysis calibration. Better put them into `storage_gas` then - it's only for the purpose of tracing anyway. And then it could have the same cost as the wasm memory expansion.

Reply:
Yeah, you're right.

Reply:
Ok, it looks like this is an _allocation_ cost. That's why we're seeing the same 6.5gas/byte cost show up in multiple places.